shell/browser/electron_browser_handler_impl.cc

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2019 Slack Technologies, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "shell/browser/electron_browser_handler_impl.h"
+
+#include <utility>
+
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/browser_thread.h"
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/browser/render_process_host.h"
+#include "mojo/public/cpp/bindings/self_owned_receiver.h"
+
+namespace electron {
+ElectronBrowserHandlerImpl::ElectronBrowserHandlerImpl(
+    content::RenderFrameHost* frame_host)
+    : render_process_id_(frame_host->GetProcess()->GetID()),
+      render_frame_id_(frame_host->GetRoutingID()),
+      weak_factory_(this) {
+  content::WebContents* web_contents =
+      content::WebContents::FromRenderFrameHost(frame_host);
+  content::WebContentsObserver::Observe(web_contents);
+}
+
+ElectronBrowserHandlerImpl::~ElectronBrowserHandlerImpl() = default;
+
+void ElectronBrowserHandlerImpl::BindTo(
+    mojo::PendingReceiver<mojom::ElectronBrowser> receiver) {
+  // Note: BindTo might be called for multiple times.
+  if (receiver_.is_bound())
+    receiver_.reset();
+
+  receiver_.Bind(std::move(receiver));
+  receiver_.set_disconnect_handler(base::BindOnce(
+      &ElectronBrowserHandlerImpl::OnConnectionError, GetWeakPtr()));
+}
+
+void ElectronBrowserHandlerImpl::WebContentsDestroyed() {
+  delete this;
+}
+
+void ElectronBrowserHandlerImpl::OnConnectionError() {
+  if (receiver_.is_bound())
+    receiver_.reset();
+}
+
+void ElectronBrowserHandlerImpl::Message(bool internal,
+                                         const std::string& channel,
+                                         blink::CloneableMessage arguments) {
+  content::RenderFrameHost* render_frame_host =
+      content::RenderFrameHost::FromID(render_process_id_, render_frame_id_);
+  api::WebContents* api_web_contents = api::WebContents::From(
+      content::WebContents::FromRenderFrameHost(render_frame_host));
```

Comment:
```suggestion
  api::WebContents* api_web_contents = api::WebContents::From(web_contents());
```

& similarly in the other methods below.

`web_contents()` comes from `content::WebContentsObserver` and will be valid always during the lifetime of this class, since we `Observe` in the constructor and `delete this` when the observed WebContents is destroyed.

Reply:
👍